File: Example_Clients/CameraClient/MicrosoftCognitiveServiceExample/MicrosoftCognitiveServiceRequests.cs
using System;
using System.Collections.Specialized;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.Web;
using System.Net;
using System.Drawing;
using System.Collections.Generic;
using Newtonsoft.Json;
using System.Windows.Forms;

namespace MicrosoftCognitiveServiceExample
{

    public class MicrosoftCognitiveServiceRequests     {

        private WebClient WebClient = new WebClient();

        public async void Analyze(string PictureFileName, string DownloadURL, string SubsciptionKey)
        {
            WebClient.DownloadFile(DownloadURL, PictureFileName);

            Bitmap mainbitmap = new Bitmap(Image.FromFile(PictureFileName));

            try
            {
                HttpResponseMessage cognitiveServiceResponse = await SendEmotionRequest(PictureFileName, SubsciptionKey);

                List<EmotionFace> emotionfaceList = JsonConvert.DeserializeObject<List<EmotionFace>>(await cognitiveServiceResponse.Content.ReadAsStringAsync());

                if (emotionfaceList != null)
                {
                    if (emotionfaceList.Count == 0)
                    {
                        UserInterface.ResultsOfAnalysis.Text = "No faces detected. \r\n";
                    }
                    else
                    {
                        for (int index = 0; index < emotionfaceList.Count; index++)
                        {
                            UserInterface.CameraSnapshot.SizeMode = PictureBoxSizeMode.CenterImage;
                            UserInterface.CameraSnapshot.Image = UserInterface.cropThisRect(mainbitmap, new Rectangle(emotionfaceList[index].faceRectangle.left, emotionfaceList[index].faceRectangle.top, emotionfaceList[index].faceRectangle.width, emotionfaceList[index].faceRectangle.height));

                            UserInterface.ResultsOfAnalysis.Text = "Results: \r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.happiness) + " happiness\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.sadness) + " sadness\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.neutral) + " neutral\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.surprise) + " surprise\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.anger) + " anger\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.contempt) + " contempt\r\n"
                            + string.Format("{0:P1}", emotionfaceList[index].scores.fear) + " fear\r\n\r\n";
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }

        public async Task<HttpResponseMessage> SendEmotionRequest(string file, string subscriptionKey)
        {
            HttpClient client = new HttpClient();
            NameValueCollection queryString = HttpUtility.ParseQueryString(string.Empty);

            // request URL
            string SERVICE_ENDPOINT = "https://westus.api.cognitive.microsoft.com/emotion/v1.0/recognize";

            // Request header
            client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", subscriptionKey);

            // Request parameter
            queryString["visualFeatures"] = "Categories";

            byte[] byteData = System.IO.File.ReadAllBytes(file);
            HttpResponseMessage response;

            // Request body / contentremote 
            using (ByteArrayContent byteArrayContent = new ByteArrayContent(byteData))
            {
                byteArrayContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                response = await client.PostAsync(SERVICE_ENDPOINT, byteArrayContent);
            }

            return response;

        }

        public async Task<HttpResponseMessage> SendAnalysisRequest(string file, string subscriptionKey)
        {
            HttpClient client = new HttpClient();
            NameValueCollection queryString = HttpUtility.ParseQueryString(string.Empty);
            client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", subscriptionKey);
            queryString["visualFeatures"] = "Tags,Description,Faces";
            string uri = "https://api.projectoxford.ai/vision/v1.0/analyze?" + (object)queryString;
            byte[] byteData = System.IO.File.ReadAllBytes(file);
            HttpResponseMessage response;
            using (ByteArrayContent byteArrayContent = new ByteArrayContent(byteData))
            {
                byteArrayContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                response = await client.PostAsync(uri, byteArrayContent);
            }

            return response;

        }

        
    }
}

File: Example_Clients/CognitiveServiceExample/src/MicrosoftCognitiveServiceRequests.java
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.net.URI;

public class MicrosoftCognitiveServiceRequests {


    public void Analyze(File picture, String subsciptionKey) {

        try {
            String cognitiveServiceResponse = this.SendEmotionRequest(picture, subsciptionKey);

            System.out.println(cognitiveServiceResponse);
            /*

            List<EmotionFace> emotionfaceList = ((JsonConvert.DeserializeObject
                    < (List < EmotionFace))
                    + cognitiveServiceResponse.Content.ReadAsStringAsync());
            if ((emotionfaceList != null)) {
                if ((emotionfaceList.Count == 0)) {
                    UserInterface.resultsOfAnalysis.Text = "No faces detected. \r\n";
                }
                else {
                    for (int index = 0; (index < emotionfaceList.Count); index++) {
                        UserInterface.CameraSnapshot.SizeMode = PictureBoxSizeMode.CenterImage;
                        UserInterface.CameraSnapshot.Image = UserInterface.cropThisRect(mainbitmap, new Rectangle(emotionfaceList[index].faceRectangle.left, emotionfaceList[index].faceRectangle.top, emotionfaceList[index].faceRectangle.width, emotionfaceList[index].faceRectangle.height));
                        UserInterface.resultsOfAnalysis.Text = ("Results: \r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.happiness) + (" happiness\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.sadness) + (" sadness\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.neutral) + (" neutral\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.surprise) + (" surprise\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.anger) + (" anger\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.contempt) + (" contempt\r\n"
                                + (String.Format("{0:P1}", emotionfaceList[index].scores.fear) + " fear\r\n\r\n"))))))))))))));
                    }

                }

            }
*/
        } catch (Exception ex) {
            ex.printStackTrace();
        }

    }

    public String SendEmotionRequest(File picture, String subscriptionKey) {

        HttpClient httpclient = HttpClients.createDefault();

        String analysis = "";

        try {
            URIBuilder builder = new URIBuilder("https://westus.api.cognitive.microsoft.com/emotion/v1.0/recognize");

            URI uri = builder.build();
            HttpPost request = new HttpPost(uri);

            boolean useJson = true;
            StringEntity reqEntity;

            if (useJson)
            {
                request.setHeader("Content-Type", "application/json");
                request.setHeader("Ocp-Apim-Subscription-Key", subscriptionKey);

                String bildURL = "\"http://tutorials.ipp.boschsecurity.com/downloads/IPP/Facedetection/Bild1.jpg\"";

                // Request body
                reqEntity = new StringEntity("{\"url\":" + bildURL + "}");
            }
            else{
                request.setHeader("Content-Type", "application/octet-stream");
                request.setHeader("Ocp-Apim-Subscription-Key", subscriptionKey);

                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

                // lokal gespeichertes Bild
                picture = new File("Resources\\Bild1.jpg");

                BufferedImage imgForAnalysis = ImageIO.read(picture);

                ImageIO.write(imgForAnalysis, "jpg", outputStream);

                // transfer the file byte-by-byte to the response object

                // funktioniert nicht :(
                reqEntity = new StringEntity(outputStream.toString());

            }

            request.setEntity(reqEntity);

            HttpResponse response = httpclient.execute(request);

            System.out.println(response.toString());

            HttpEntity entity = response.getEntity();

            if (entity != null) {
                analysis = EntityUtils.toString(entity);
            }

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        return analysis;
    }

    public String SendAnalysisRequest(String file, String subscriptionKey) {

        HttpClient httpclient = HttpClients.createDefault();

        String analysis = "";

        try {
            URIBuilder builder = new URIBuilder("https://westus.api.cognitive.microsoft.com/vision/v1.0/analyze");

            builder.setParameter("visualFeatures", "Categories");
            builder.setParameter("details", "{string}");
            builder.setParameter("language", "en");

            URI uri = builder.build();
            HttpPost request = new HttpPost(uri);
            request.setHeader("Content-Type", "application/json");
            request.setHeader("Ocp-Apim-Subscription-Key", "{subscription key}");


            // Request body
            StringEntity reqEntity = new StringEntity("{body}");
            request.setEntity(reqEntity);

            HttpResponse response = httpclient.execute(request);
            HttpEntity entity = response.getEntity();

            if (entity != null) {
                analysis = EntityUtils.toString(entity);
            }

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        return analysis;
    }
}